src/storage/data_table.cpp

```diff
@@ -314,15 +359,13 @@ bool DataTable::CompareAndSwapVersionPtr(const TupleSlot slot, const TupleAccess
   return reinterpret_cast<std::atomic<UndoRecord *> *>(ptr_location)->compare_exchange_strong(expected, desired);
 }
 
-void DataTable::NewBlock(RawBlock *expected_val) {
+std::list<RawBlock *>::iterator DataTable::NewBlock() {
   common::SpinLatch::ScopedSpinLatch guard(&blocks_latch_);
-  // Want to stop early if another thread is already getting a new block
-  if (expected_val != insertion_head_) return;
   RawBlock *new_block = block_store_->Get();
   accessor_.InitializeRawBlock(this, new_block, layout_version_);
```

Comment:
Move these two lines to before the instantiation of the spin latch.  It should greatly reduce the time spent in the critical section without affecting correctness.